src/mono/mono/metadata/monitor.c

```diff
@@ -1430,6 +1430,29 @@ mono_monitor_wait (MonoObjectHandle obj_handle, guint32 ms, MonoBoolean allow_in
 	return success;
 }
 
+#ifdef HOST_WASM
+void
+mono_set_string_interned_internal (MonoObject* obj)
+{
+	LockWord tmp;
+	tmp.sync = obj->synchronisation;
+	if (!lock_word_is_inflated (tmp))
+		mono_monitor_inflate (obj);
```

Comment:
@kg I think this might be the cause of the ` Object synchronization method was called from an unsynchronized block of code.` exception.  `mono_monitor_inflate` I think doesn't preserve the "owner" if the string is already used as a lock.

So something like:
```
  string s = "123";
  Monitor.Enter (s);
  Intern(ref s); // or whatever is the indirect operation that would intern s
  Monitor.Exit (s); // oops - interning lost the "owner"
```

Reply:
I mentioned this in chat but I'll mention it here for posterity: While you're correct about this I can't find a way to actually produce the outcome you describe, because our interning implementation always makes a pinned copy of the input string first - the return value of String.Intern in mono seems to *never* be the input string, unless the input string was already interned.

As such I couldn't come up with a test to verify whether we maintain the owner and other state.
Just for the sake of completeness I made an attempt at preserving it anyway, and didn't observe any new failures produced by it... but I'm not sure whether it would actually address this problem.

Reply:
This new code is definitely broken because it updates the status on the inflated lock without any CAS. Also, I don't understand why you say inflating the monitor loses the owner. The inflating operation takes the current monitor owner from the LockWork here https://github.com/dotnet/runtime/blob/main/src/mono/mono/metadata/monitor.c#L505

Reply:
Thanks, I'll remove this commit then.

Reply:
Yea, I was wrong.  looks like it doesn't lose the owner.